src/actions/toggleNote.ts

```diff
@@ -1,35 +1,57 @@
-import findDescendant from '../selectors/findDescendant'
 import getChildren from '../selectors/getChildren'
+import getThoughtById from '../selectors/getThoughtById'
+import resolveNotePath from '../selectors/resolveNotePath'
 import { registerActionMetadata } from '../util/actionMetadata.registry'
 import head from '../util/head'
 import reducerFlow from '../util/reducerFlow'
+import deleteAttribute from './deleteAttribute'
+import deleteThought from './deleteThought'
+
+/**
+ * Calculate the offset for note focus positioning.
+ */
+const calculateOffset = (state: State, thoughtId: ThoughtId | undefined): number => {
+  if (!thoughtId) return 0
+  const noteChildren = getChildren(state, thoughtId)
+  return noteChildren.length ? noteChildren[0].value.length : 0
+}
 
 /** Toggle the caret between the cursor and its note. Set the selection to the end. If the note is empty, delete it. */
 const toggleNote = (state: State): State => {
-  const thoughtId = head(state.cursor!)
-  const hasNote = findDescendant(state, thoughtId, '=note')
-  const noteChildren = hasNote ? getChildren(state, hasNote) : []
-  const offset = noteChildren.length ? noteChildren[0].value.length : 0
+  const path = state.cursor!
+  const targetThoughtPath = resolveNotePath(state, path)
+  const targetThought = targetThoughtPath ? getThoughtById(state, head(targetThoughtPath)) : undefined
+  const offset = calculateOffset(state, targetThought?.id)
+  const isNoteContentEmpty = !offset
+  const isNoteAttribute = targetThought && targetThought.value === '=note'
 
   return reducerFlow([
-    // create an empty note if it doesn't exist
-    !hasNote
-      ? setDescendant({
-          path: state.cursor!,
-          values: ['=note', ''],
-        })
-      : // delete an empty note that already exists
-        state.noteFocus && !attribute(state, thoughtId, '=note')
-        ? deleteAttribute({ path: state.cursor!, value: '=note' })
+    // Create note if target thought doesn't exist (=note attribute)
+    !targetThought
+      ? state => setDescendant(state, { path: state.cursor!, values: ['=note', ''] })
+      : !isNoteAttribute && isNoteContentEmpty
+        ? state.noteFocus
+          ? // delete the target thought if it exists via =path
+            state =>
+              deleteThought(state, {
+                pathParent: path,
+                thoughtId: targetThought.id,
+              })
+          : // create a missing target thought if it doesn't exist via =path
+            state => setDescendant(state, { path: state.cursor!, values: [targetThought?.value, ''] })
         : null,
 
-    // toggle state.noteFocus, which will trigger the Editable and Note to re-render and set the selection appropriately
+    // if targetThought is a noteAttribute and the note content is empty and noteFocus is true, delete the note attribute
+    state.noteFocus && isNoteAttribute && isNoteContentEmpty
+      ? deleteAttribute({ path: state.cursor!, value: '=note' })
```

Comment:
This `deleteAttribute` is far apart from `deleteThought` above, even though they represent the same delete action when toggling a note off. Structuring the logic so these are grouped together would probably make more sense.